### PR TITLE
Add property buildLabels to add build info to image

### PR DESCRIPTION
**Pull request by llaforest**
#816 
Adding labels into this property will not invalidate the cache key.
It might be useful to add manifest information without trigerring
a new build every time.
But when a build is triggered, the information goes in.

### CODE REVIEW

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 1]
```diff
@@ -87,6 +87,9 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
     @Optional
     final MapProperty<String, String> labels = project.objects.mapProperty(String, String)
 
+    @Optional
```
Comment (hugares) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Shouldn't this property be annotated with @Internal to indicate that it is neither an input nor an output.

Comment (llaforest) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Yes I think it should be annotated with @Internal

**Comment by llaforest** [comment]
Any comments on this feature?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 2]
```diff
@@ -87,6 +87,14 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
     @Optional
     final MapProperty<String, String> labels = project.objects.mapProperty(String, String)
 
+    /**
+     * The buildLabels are labels that will not invalidate the cache key when changed.
+     * Usually used to store in the image build information that changes all the time.
+     * They will only be applied when the image builds, otherwise they are ignored.
+     */
+    @Internal
+    final MapProperty<String, String> buildLabels = project.objects.mapProperty(String, String)
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: I'm not sure how else to address this but it _could_ become confusing as folks may not know they can actually use it given the annotation.

Comment (cdancy) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: Can you **also** add an @input and @optional along with keeping the @internal ?

Comment (llaforest) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: The goal of this feature is 'not to be part of the inputs'
Let say you need some references in your image about the date it was build, the git id of the commit you used, stuff like that will change all the time and will invalidate the cache key all the time if part of the input.
My claim is that we need a property that enables us to save some information in the image when it is rebuilt, but it must not trigger the rebuild.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy [hunk 3]
```diff
@@ -382,6 +422,27 @@ class DockerBuildImageFunctionalTest extends AbstractGroovyDslFunctionalTest {
...
+            }
+
+            task buildImage(type: DockerBuildImage) {
+                dependsOn dockerfile
+            }
+
+            task inspectImage(type: DockerInspectImage) {
+                dependsOn buildImage
+                targetImageId buildImage.getImageId()
+            }
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy]: Can you add a cleanup task, which will remove the image?

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 4]
```diff
@@ -192,7 +201,11 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
         }
 
         if (labels.getOrNull()) {
-            buildImageCmd.withLabels(labels.get().collectEntries { [it.key, it.value.toString()] })
+            buildLabels.putAll(labels.get())
```
Comment (orzeh) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: So `labels` can silently override `buildLabels`? This may cause some nasty bugs I guess.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 5]
```diff
@@ -142,6 +150,7 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
             if(file.exists()) {
                 try {
                     getDockerClient().inspectImageCmd(file.text).exec()
+                    imageId.set(file.text)
```
Comment (orzeh) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: This change is unrelated to the scope of this PR.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy (hunk 2 above)
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: A task property can only be ignored for up-to-date checking by not annotating it or annotating it with `@Internal`. It doesn't make sense to combine `@Internal` with any other input or output annotation.

File: src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy (hunk 3 above)
Comment (llaforest) [reply src/functTest/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImageFunctionalTest.groovy]: I could but non of the test cleanup so far. I was more thinking of implementing a cleanup for the whole test case.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy (hunk 5 above)
Comment (llaforest) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: I agree it could be moved to another issue

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy (hunk 4 above)
Comment (llaforest) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: No they will add to each other. They are complementary. If some labels are really part of the build image inputs, then they should be 'labels'. If they are just complementary informations, they should be 'buildLabels'
